examples/name-service-without-macros/programs/name-service/src/lib.rs

```diff
@@ -0,0 +1,219 @@
+use std::net::{Ipv4Addr, Ipv6Addr};
+
+use anchor_lang::prelude::*;
+use borsh::{BorshDeserialize, BorshSerialize};
+use light_hasher::bytes::AsByteVec;
+use light_sdk::{
+    compressed_account::{LightAccount, LightAccounts, OutputCompressedAccountWithPackedContext},
+    context::LightInstructionInputs,
+    light_system_accounts, LightDiscriminator, LightHasher, LightTraits,
+};
+
+declare_id!("7yucc7fL3JGbyMwg4neUaenNSdySS39hbAk89Ao3t1Hz");
+
+#[program]
+pub mod name_service {
+    use light_sdk::context::LightContext;
+
+    use super::*;
+
+    pub fn create_record<'info>(
+        ctx: Context<'_, '_, '_, 'info, CreateRecord<'info>>,
+        inputs: Vec<u8>,
+    ) -> Result<()> {
+        let inputs = CreateRecordInputs::deserialize(&mut inputs.as_slice())?;
+        let mut ctx =
+            LightContext::<CreateRecord, LightCreateRecord>::new(ctx, inputs.light_inputs)?;
+
+        ctx.light_accounts.record.owner = ctx.accounts.signer.key();
+        ctx.light_accounts.record.name = inputs.name;
+        ctx.light_accounts.record.rdata = inputs.rdata;
+
+        ctx.verify()?;
```

Comment:
An abstraction level that seems intuitive to be is:
```
pub fn create_record<'info>(
        ctx: Context<'_, '_, '_, 'info, CreateRecord<'info>>,
        inputs: Vec<u8>,
    ) -> Result<()> {
       let mut start = 0;
       let proof = Proof::deserialize(&mut start, inputs);
       let record = CreateRecordInputs::deserialize(&mut start, inputs);
       let new_addresses = NewAddresses::deserialize(&mut start, inputs);
       
        record.owner = ctx.accounts.signer.key();
        record.name = inputs.name;
        record.rdata = inputs.rdata;
        
        verify(ctx.accounts,  Some(proof),Some(vec![record.into()]), Some(new_addresses))?;
 } 

pub fn verify<T: LightAccountsTrait>(
  accounts: T, 
  proof: Option<ProofRpcResult>,
  compressed_accounts: Option<Vec<PackedCompressedAccountWithMerkleContext>>,
  new_addresses: Option<Vec<PackedNewAddressParams>>
) -> Result<(), SdkError>

In an anchor program which uses `#[light_program]` we can push `manually` (its not manual since the macro generates the code) into an account vector if that makes sense.

Reply:
I found a solution which wouldn't even require having an offset variable! There is [`BorshDeserialize::deserialize_reader`](https://docs.rs/borsh/0.10.4/borsh/de/trait.BorshDeserialize.html#tymethod.deserialize_reader) which is [used internally by other methods](https://docs.rs/borsh/0.10.4/src/borsh/de/mod.rs.html#39).

This example:

https://github.com/vadorovsky/anchor-borsh-experiments/blob/main/borsh-playground/src/main.rs

Proves that reading multiple serialized structs with combination of [`std::io::Cursor`]() (which is just a dummy wrapper around any buffer with a counter). The output is:

```
buf [serialized: foo]: [69, 0, 0, 0, 3, 0, 0, 0, 102, 111, 111]
buf [serialized: foo, bar]: [69, 0, 0, 0, 3, 0, 0, 0, 102, 111, 111, 3, 0, 0, 0, 98, 97, 114, 164, 1, 0, 0]
foo (deserialized): Foo { a: 69, b: "foo" }
bar [serialized: bar] [deserialized: foo]: Cursor { inner: [69, 0, 0, 0, 3, 0, 0, 0, 102, 111, 111, 3, 0, 0, 0, 98, 97, 114, 164, 1, 0, 0], pos: 11 }
bar (deserialized): Bar { a: "bar", b: 420 }
bar [deserialized: foo, bar]: Cursor { inner: [69, 0, 0, 0, 3, 0, 0, 0, 102, 111, 111, 3, 0, 0, 0, 98, 97, 114, 164, 1, 0, 0], pos: 22 }
```

So I think your code could be changed to something like

```rust
pub fn create_record<'info>(
        ctx: Context<'_, '_, '_, 'info, CreateRecord<'info>>,
        inputs: Vec<u8>,
    ) -> Result<()> {
       let mut inputs = Cursor::new(inputs);
       let user_inputs = UserInputs::deserialize_reader(&mut inputs);
       let proof = Proof::deserialize_reader(&mut inputs);
       let new_addresses = PackedAddressMerkleContexts::deserialize_reader(&mut inputs);
       let (new_address_params, addresses) = check_address(&[inputs.name], new_addresses);
       let record = CreateRecordInputs::deserialize_init(&mut inputs, Some(addresses[0]));
       
        record.owner = ctx.accounts.signer.key();
        record.name = inputs.name;
        record.rdata = inputs.rdata;
        
        verify(ctx.accounts,  Some(proof),Some(vec![record.into()]), Some(new_address_params))?;
 } 
```

Reply:
that makes sense

Reply:
we should measure CU usage of this though

Reply:
Yes, happy to do that (maybe even today / tomorrow morning).